lib/ansible/modules/system/vdo.py

```diff
@@ -0,0 +1,762 @@
+#!/usr/bin/python
+
+#
+# Copyright (c) 2017 Red Hat, Inc.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.
+#
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {
+    'metadata_version': '1.1',
+    'status': ['preview'],
+    'supported_by': 'community'
+}
+
+DOCUMENTATION = '''
+---
+author:
+    - "Bryan Gurney (@bgurney-rh)"
+
+module: vdo
+
+short_description: Module to control VDO
+
+version_added: "2.5"
+
+description:
+    - "This module controls the VDO dedupe and compression device."
+
+options:
+    name:
+        description:
+            - The name of the VDO volume.
+        required: true
+    state:
+        choices: [ "present", "absent" ]
+        description:
+            - Whether this VDO volume should be "present" or "absent".
+              If a "present" VDO volume already exists, it will not be
+              created. If an "absent" VDO volume does not exist, it
+              will not be removed.
```

Comment:
I'll update this section.  If a VDO volume already exists, it will not be created, but if the differences between the current state and the playbook parameters are conducive to doing so, this module will modify the VDO volume.  Some modifications (start, stop, and growphysical if the playbook contains "growphysical: yes") will take effect at playbook execution time.

However, since the module needs to perform a match between the currently running settings and the settings in the playbook, several of the options that appear to be boolean cannot actually be made boolean, since they must match the output strings used by the "vdo status" command.  I think the only option that could presently be made boolean is "activated".

Reply:
the UI does not have to be a direct passthrough, its preferable to leave these as booleans as it makes it simpler for the user to use 'jinja2 expressions'

Reply:
I don't believe you understood my reply: if the playbook options do not exactly match with the output format of the "vdo status" command, change detection will fail.  I am making a conscious decision here for the playbook UI to be a direct passthrough of the "VDO" command.  The choice options detailed below are a reflection of that conscious decision.

The "activated" playbook parameter is a compromise; that has a shim to act as a boolean.

Reply:
...let me see if I can allow booleans for those playbook options, and internally translate them to the proper strings...

Reply:
There's a complication here: for Ansible 2.3.2 (the version of Ansible being used for testing), "type='bool'" results in "implementation error: unknown type <type 'bool'>."